### PR TITLE
[receiver/googlecloudmonitoringreceiver] Adding Google Cloud monitoring receiver

### PR DESCRIPTION
**Pull request by abhishek-at-cloudwerx**
**Description:** Adding Google Cloud Monitoring receiver

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33762

**Testing:** Test cases for configuration added

**Documentation:** README file added for describing the component

### CODE REVIEW

**Comment by crobert-1** [comment]
Hello @abhishek-at-cloudwerx, please sign the CLA to enable this PR to be reviewed.

**Comment by abhishek-at-cloudwerx** [comment]
@crobert-1 - Yes we're working on getting the org CLA sorted. Unfortunately we are unable to get any help in figuring out who registered as Org admin.

**Comment by abhishek-at-cloudwerx** [comment]
@crobert-1 - The CLA is now signed and PR is ready for review. 

@dashpole - Tagging as Sponsor 
@TylerHelmuth - to be one of the CODEOWNERS

File: receiver/googlecloudmonitoringreceiver/README.md
Comment (dashpole) [inline receiver/googlecloudmonitoringreceiver/README.md:42]: please document `initial_delay` and `timeout` options from scaperhelper as well

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (dashpole) [inline receiver/googlecloudmonitoringreceiver/config.go:19]: What is the region going to be used for?  Are you going to call a regional endpoint, or are you using this as part of the query filter?

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (dashpole) [inline receiver/googlecloudmonitoringreceiver/config.go:28]: This seems to duplicate functionality from `initial_delay` in scraperhelper.ControllerConfig.  Consider removing.

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (dashpole) [inline receiver/googlecloudmonitoringreceiver/config.go:22]: A Service, as defined below, only has a single metric.  I would consider renaming this `metrics`.

File: receiver/googlecloudmonitoringreceiver/README.md
Comment (dashpole) [inline receiver/googlecloudmonitoringreceiver/README.md:54]: Did you forget to remove this section?  It looks like you aren't exposing the raw filter to users.

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/config.go:19]: Since it's optional, removing it from the config.

File: receiver/googlecloudmonitoringreceiver/README.md
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/README.md:54]: No, metric_name should be included in the raw filter, which is why I categorised it under title filtering.

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/config.go:28]: We are using individual delay because certain services cannot display time series data immediately.
For exa- Read the **compute** service parameters [](https://cloud.google.com/monitoring/api/metrics_gcp#gcp-compute)

**Comment by abhishek-at-cloudwerx** [comment]
@dashpole I made the changes as per your comment and removed the **service_name** field since we now fetch details directly based on the **metric_name**.

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/config.go]: Updated

File: cmd/otelcontribcol/go.mod
Comment (dashpole) [inline cmd/otelcontribcol/go.mod:5]:
```suggestion
go 1.21.0

toolchain go1.21.12
```

File: cmd/oteltestbedcol/go.mod
Comment (dashpole) [inline cmd/oteltestbedcol/go.mod:7]:
```suggestion
toolchain go1.21.12
```

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (dashpole) [reply receiver/googlecloudmonitoringreceiver/config.go]: Does the list request return an error, or just no metrics? It might be easier to just not return metrics instead of requiring users to configure a delay.

File: receiver/googlecloudmonitoringreceiver/README.md
Comment (dashpole) [inline receiver/googlecloudmonitoringreceiver/README.md:51]: I would recommend linking to https://cloud.google.com/docs/authentication/application-default-credentials.  Note that workloads on GCP will use the service accounts credentials without setting the environment variable.

**Review by dashpole — APPROVED** [review]
lgtm after nits addressed

File: receiver/googlecloudmonitoringreceiver/README.md
Comment (atoulme) [inline receiver/googlecloudmonitoringreceiver/README.md:17]:
```suggestion
This receiver gets GCP (Google Cloud Platform) metrics from [GCP Monitoring REST API] via the [Google SDK for GCP Metrics] and then convert those timeseries data into OTel Format [Pipeline Data].
```

**Review by atoulme — APPROVED** [review]
LGTM after nits addressed, @TylerHelmuth should review as well

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/config.go]: If the data is unavailable, a "no more items in iterator" error will be displayed.

To obtain time-series data based on delay configuration, we need to apply individual delays as specified. Otherwise, a global delay will suffice.

File: receiver/googlecloudmonitoringreceiver/README.md
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/README.md]: Updated

File: receiver/googlecloudmonitoringreceiver/README.md
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/README.md]: Updated

File: cmd/oteltestbedcol/go.mod
Comment (abhishek-at-cloudwerx) [reply cmd/oteltestbedcol/go.mod]: Done

File: cmd/otelcontribcol/go.mod
Comment (abhishek-at-cloudwerx) [reply cmd/otelcontribcol/go.mod]: Done

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (dashpole) [reply receiver/googlecloudmonitoringreceiver/config.go]: sounds reasonable for now.

**Review by shrutimantri — APPROVED** [review]

**Comment by abhishek-at-cloudwerx** [comment]
@TylerHelmuth @dashpole - Could one of you help merge this ? 
Thanks

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (TylerHelmuth) [inline receiver/googlecloudmonitoringreceiver/config.go:51]: Is a value of 0 ok?

**Review by TylerHelmuth — COMMENTED** [review]
Please take a look at the failing CI

File: receiver/googlecloudmonitoringreceiver/config.go
Comment (abhishek-at-cloudwerx) [reply receiver/googlecloudmonitoringreceiver/config.go:51]: No, this validation is intended to reject negative or zero values.

**Review by TylerHelmuth — APPROVED** [review]

**Comment by abhishek-at-cloudwerx** [comment]
Thanks @TylerHelmuth for approval. @djaglowski could you please help with the final approval and merge as CODEOWNER.

**Review by djaglowski — COMMENTED** [review]
@abhishek-at-cloudwerx LGTM but I'm not a codeowner on this component.

**Comment by dashpole** [comment]
Error: googlecloudmonitoringreceiver/receiver.go:28:37: unused-parameter: parameter 'ctx' seems to be unused, consider removing or renaming it as _ (revive)
func (m *monitoringReceiver) Scrape(ctx context.Context) (pmetric.Metrics, error) {
                                    ^
Error: googlecloudmonitoringreceiver/receiver.go:49:41: unused-parameter: parameter 'ctx' seems to be unused, consider removing or renaming it as _ (revive)
func (m *monitoringReceiver) initialize(ctx context.Context) error {
                                        ^

**Comment by abhishek-at-cloudwerx** [comment]
@dashpole - For now, added a temporary use of the variable to ensure the warning doesn't show up. Since this is a skeleton code, the variable is currently unused. In the next PR, this is expected to be used.

**Comment by dashpole** [comment]
Looks like you need to update collector core dependencies in go.mod after the rebase

**Comment by dashpole** [comment]
Still some presubmit failures

**Comment by abhishek-at-cloudwerx** [comment]
> Still some presubmit failures

@dashpole - Yes we're currently reviewing those and will drop a note once all is in good shape.

**Pull request merged**
